osquery/tables/events/windows/powershell_events.cpp

```diff
@@ -6,212 +6,355 @@
+#include <boost/filesystem/path.hpp>
 
-#include <boost/property_tree/ptree.hpp>
-#include <boost/property_tree/xml_parser.hpp>
-
-#include <osquery/config/config.h>
-#include <osquery/core.h>
 #include <osquery/database.h>
+#include <osquery/flags.h>
 #include <osquery/logger.h>
 #include <osquery/registry_factory.h>
-#include <osquery/tables.h>
-
-#include <osquery/core/windows/wmi.h>
-#include <osquery/events/windows/windows_event_log.h>
-#include <osquery/filesystem/fileops.h>
+#include <osquery/tables/events/windows/powershell_events.h>
 #include <osquery/utils/conversions/tryto.h>
-#include <osquery/utils/json/json.h>
-#include <plugins/config/parsers/feature_vectors.h>
-
-namespace pt = boost::property_tree;
 
 namespace osquery {
+namespace {
+const std::time_t kScriptExpirationSchedule{120};
+const std::time_t kScriptEventExpiration{60};
 
-const std::string kPowershellDomain{"powershell_script_block_logs"};
 const std::string kScriptBlockPrefix{"script_block."};
-const std::string kScriptBlocksReceivedSuffix{".blocks_received"};
-const std::wstring kPowershellEventsChannel{
-    L"microsoft-windows-powershell/operational"};
+const std::string kPowershellEventsChannel{
+    "microsoft-windows-powershell/operational"};
 
 const int kScriptBlockLoggingEid{4104};
-const int kCharFreqVectorLen{256};
-
-void parseTree(const pt::ptree& tree, std::map<std::string, std::string>& res);
-
-class PowershellEventSubscriber
-    : public EventSubscriber<WindowsEventLogEventPublisher> {
- public:
-  Status init() override {
-    // Before starting our subscription, purge any residual db entries as it's
-    // unlikely we'll finish re-assembling them
-    std::vector<std::string> keys;
-    scanDatabaseKeys(kEvents, keys, kScriptBlockPrefix);
-    for (const auto& k : keys) {
-      auto s = deleteDatabaseValue(kEvents, k);
-      if (!s.ok()) {
-        VLOG(1) << "Failed to delete stale script block from the database "
-                << k;
-      }
-    }
 
-    auto wc = createSubscriptionContext();
-    wc->sources.insert(kPowershellEventsChannel);
+void cleanupLegacyDatabaseEntries() {
+  std::vector<std::string> key_list;
+  scanDatabaseKeys(kEvents, key_list, kScriptBlockPrefix);
 
-    subscribe(&PowershellEventSubscriber::Callback, wc);
-    return Status::success();
+  bool display_error{false};
+  for (const auto& key : key_list) {
+    auto status = deleteDatabaseValue(kEvents, key);
+    if (!status.ok()) {
+      display_error = true;
+    }
   }
 
-  Status Callback(const ECRef& ec, const SCRef& sc);
+  if (display_error) {
+    LOG(ERROR) << "Failed to delete stale script blocks from the database";
+  }
+}
+} // namespace
 
-  void addScriptResult(Row& results);
-};
+FLAG(bool, enable_powershell_events, false, "Enables Powershell events");
+DECLARE_bool(enable_windows_events);
 
 REGISTER(PowershellEventSubscriber, "event_subscriber", "powershell_events");
 
-inline double cosineSimilarity(std::vector<double>& script_freqs,
```

Comment:
The character frequency was previously read and parsed any time a new script result was being added by the powershell_events table.

In the refactor, this configuration is only read and parsed once whenever the configuration is reloaded by the publisher. This setting is then passed down to subscribers that may wish to use it later on.

The reason the `cosineSimilarity` function has been moved is to have them both in the sample place rather than splitting the configuration reader/parser in the publisher and the configuration user inside one of the subscribers